libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationService.java

```diff
@@ -261,32 +265,26 @@ public void onUserOffRoute(Location location, boolean userOffRoute) {
 
   public void rerouteOccurred() {
     mapboxNavigation.setSessionState(mapboxNavigation.getSessionState().toBuilder()
+      .routeProgressBeforeReroute(routeProgress)
       .beforeRerouteLocations(Arrays.asList(
         locationBuffer.toArray(new Location[locationBuffer.size()])))
-      .routeProgressBeforeReroute(routeProgress)
+      .lastRerouteDate(new Date())
       .build());
-    locationBuffer.clear();
-    queuedRerouteEvent = true;
-
-    handler = new Handler();
-    runnable = new Runnable() {
-      @Override
-      public void run() {
-        mapboxNavigation.setSessionState(mapboxNavigation.getSessionState().toBuilder()
-          .afterRerouteLocations(Arrays.asList(
-            locationBuffer.toArray(new Location[locationBuffer.size()])))
-          .build());
-
-        locationBuffer.clear();
-
-        NavigationMetricsWrapper.rerouteEvent(mapboxNavigation.getSessionState(), routeProgress,
-          location);
-        queuedRerouteEvent = false;
-      }
-    };
-    handler.postDelayed(runnable, TWENTY_SECOND_INTERVAL);
+
+    queuedRerouteEvents.add(mapboxNavigation.getSessionState());
   }
 
+  void sendRerouteEvent(SessionState sessionState) {
+    mapboxNavigation.setSessionState(sessionState.toBuilder()
```

Comment:
By capturing the session state at the time of reroute, how do we capture and send up the new geometry, etc associated resulting from the reroute?

Reply:
iOS updates this info here after the reroute request comes back: 

https://github.com/mapbox/mapbox-navigation-ios/blob/41860934843653fd6f7402219e6bb179482f71bd/MapboxCoreNavigation/RouteController.swift#L405-L410

https://github.com/mapbox/mapbox-navigation-ios/blob/41860934843653fd6f7402219e6bb179482f71bd/MapboxCoreNavigation/MMEEventsManager.swift#L354-L362